enumerable_methods.rb

```diff
@@ -0,0 +1,106 @@
+module Enumerable
+  # 1. my_each method
+  def my_each
+    size.times { |counter| yield(to_a[counter]) } if block_given?
+    block_given? ? self : enum_for(__method__)
+  end
+
+  # 2. my each with index methods
+  def my_each_with_index
+    enum = to_a
+    return enum_for unless block_given?
+
+    enum.length.times do |arr_index|
+      yield(enum[arr_index], arr_index)
+    end
+    self
+  end
+
+  # 3. my select method
+  def my_select
+    result = []
+    return to_enum(:my_select) unless block_given?
+
+    my_each do |arr_item|
+      result << arr_item if yield(arr_item)
+    end
+    result
+  end
+
+  # 4. my all method
+  def my_all?(*args)
+    result = true
+    if !args[0].nil?
+      my_each { |element| result = false unless args[0] === element }
+    elsif !block_given?
+      my_each { |element| result = false unless element }
+    else
+      my_each { |element| result = false unless yield(element) }
+    end
+    result
+  end
+
+  # 5. my any method
+  def my_any?(*arg)
+    result = false
+    if !arg[0].nil?
+      my_each { |element| result = true if arg[0] === element }
+    elsif !block_given?
+      my_each { |element| result = true if element }
+    else
+      my_each { |element| result = true if yield(element) }
+    end
+    result
+  end
+
+  # 6. my none method
+  def my_none?(arg = nil, &block)
+    !my_any?(arg, &block)
+  end
+
+  # 7. my count method
+  def my_count(argument = nil)
+    result = []
+    count = 0
+    counter = proc { |_item, index| count = index + 1 }
+    if argument
+      my_each { |item| result << item if argument == item }
+      result.my_each_with_index(&counter)
+    elsif block_given?
+      my_each { |item| result << item if yield(item) }
+      result.my_each_with_index(&counter)
+    else
+      my_each_with_index(&counter)
+    end
+    is_a?(Hash) ? count - 2 : count
+  end
+
+  # 8. my map method
+  def my_map
```

Comment:
### The `my_map` method should only execute the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

An argument error is raised when I pass an argument to the enumerable method.
```
ArgumentError:
       wrong number of arguments (given 1, expected 0)

```

Reply:
Thanks for the feedback. We implemented the suggested changes.